### DISCUSSION TITLE
OCI Image for Policy Bundle

### DISCUSSION THREAD

**Discussion by Ronnie-personal** [open-policy-agent/community#348] [OPA and Rego] [answered]
OPA can now consume policy bundles packaged as OCI images. Has this approach been commonly adopted by any users in the production environment?

I'm just wondering if this approach will become the standard way for OPA users.
What are the benefits of using OCI images for OPA bundles?

**Comment by orweis** [comment]
Personally I don't see the appeal.
Policies from my experience are dynamic living things - that's why we opted for the [OPAL](https://github.com/permitio/opal/) approach , which allows using Git as the source of truth for policies (even directly) ; with commit hashes being your auditable policy identity.

**Reply by Ronnie-personal** [reply]
Can you please elaborate more about 'using git as the source truth for policies (even directly)'?
Do you imply two options, one is 'as the source truth' and another option 'as directly'?

**Reply by orweis** [reply]
Sure,
OPAL can track a git repository directly - i.e. changes in the git repo (and tracked branch) immediately propogate to relevant OPA agents according to the directories they subscribed to.
https://docs.opal.ac/tutorials/track_a_git_repo

OPAL can also track regular bundle servers, which, I'd still recommend will get their policies from a git repository. But with OPAL, OPA doesn't have to poll on them, but can instead be event driven, using a webhook.
https://docs.opal.ac/tutorials/track_an_api_bundle_server

**Reply by anderseknert** [reply]
@Ronnie-personal including git revision in a bundle is simple:

```shell
opa build --bundle --revision $(git rev-parse HEAD) .
```

This revision will be stored in the `.metadata` file inside the bundle, and will be included in all decision logs emitted by OPA.

**Reply by Ronnie-personal** [reply]
What are the advantages to create bundle? Can we deploy each individual policy file?

Could you please provide document links regarding OPA bundle versions and metadata? Are they the same thing or not?
Can I use either CLI or API to retrieve the bundle version and the revision metadata?

Thank you so much!

**Reply by anderseknert** [reply]
The [docs on bundles](https://www.openpolicyagent.org/docs/latest/management-bundles/) should hopefully answer that. Let me know if there's something you feel is missing in those, and we'll have them updated :)

**Reply by Ronnie-personal** [reply]
Thank you @anderseknert for sharing the document. I have reviewed the document.

I would like to understand if the bundle has a 'version' attribute besides 'revision'. Is ETAG the version of the bundle file? What value does ETAG carry?

**Reply by anderseknert** [reply]
You can use the metadata atribute if you want to provide a custom versioning scheme:
> metadata - An optional key that contains arbitrary metadata to accompany the bundle. This metadata is available for querying using data.system, along with the rest of the manifest.

The Etag is created by the bundle server, so you'll need to check your bundle server documentation for how they calculate that, but most commonly it'll be a SHA-sum of the file, I believe.

**Comment by anderseknert** [comment, marked as answer]
Hi again @Ronnie-personal 👋

I don’t believe there will be a single standard approach for distributing policy and data to OPA. If you’re using OCI registries for other things already, that may be a benefit of the distribution format for you. It’s not going to replace any other distribution mechanism though.
